### DISCUSSION TITLE
[meta-mender-tegra] self update single copy machinism

### DISCUSSION THREAD

**Discussion by EasyIP2023** [OE4T/meta-tegra#732] [Q&A]
I've briefly looked into to mender and was wondering if it is possible to do single copy updates like with swupdate? (https://sbabic.github.io/swupdate/overview.html#single-copy). Would prefer this update approach as it helps with reduction of the overall image size. Also, would rather not do dual copy A/B partition approach. Per looking into swupdate doesn't seem to work with nvidia's L4T C-boot for xavier series boards. If single-copy update approach isn't possible would it be possible to switch from c-boot over to u-boot? Or is nvidia not having it like is say in this forum post https://forums.developer.nvidia.com/t/custom-u-boot-on-xavier-solved/65885/2

**Comment by kekiefer** [comment]
Conceivably, one could boot into the recovery mode kernel + initrd to prepare the system for an upgrade. As for the feasibility of doing this with Mender, someone else would have to say. I've only seen full A/B and delta update done with mender.

Support for t19x u-boot does not exist, and sounds unlikely considering NVIDIA's recent moves towards EFI boot schemes for mainline support instead.

**Reply by EasyIP2023** [reply]
Thanks

**Comment by ichergui** [comment]
@EasyIP2023 

`SWUpdate` requires `u-boot`, maybe some image types that are not support yet such as `WIC` image. 

Nvidia `Tegra194` SOC doesn't support `u-boot`, I believe that is difficult to adapt software update mechanism when using Nvidia bootloader `cboot`.

Also, `Kernel` and `DTB` are installed in different partition, and the complicated part is when secure boot is enable (images are signed). I don't know how `SWUpdate` manage this.   

As @kekiefer mentioned, maybe [delta update](https://docs.mender.io/system-updates-yocto-project/customize-mender/delta-update-support) with mender will be interesting for you.

**Reply by kekiefer** [reply]
> SWUpdate requires u-boot, maybe some image types that are not support yet such as WIC image.

To be clear, I think you mean to say that the swupdate single copy mechanism requires u-boot, which may be true. A/B update with swupdate and cboot is no problem.

**Reply by EasyIP2023** [reply]
Thanks guys

**Reply by ichergui** [reply]
Please refer to the [link1](https://sbabic.github.io/swupdate/building-with-yocto.html) and [link2](https://github.com/sbabic/meta-swupdate/blob/master/recipes-extended/images/swupdate-image.inc#L19)

**Reply by ichergui** [reply]
@EasyIP2023 

You're welcome

**Comment by dwalkes** [comment]
One comment I wanted to add about this:

> If single-copy update approach isn't possible would it be possible to switch from c-boot over to u-boot?

I'm not sure if you've noticed this but both c-boot and u-boot are [going the way of the dodo bird with Jetpack 5](https://forums.developer.nvidia.com/t/jetson-software-roadmap-for-2h-2021-and-2022/177721), so it would be better to end up with something which was as decoupled as possible from either.   Also, with u-boot you can end up with multiple sources of truth for rootfs which can be confusing and difficult to manage.  See discussion at https://github.com/BoulderAI/meta-mender-community/pull/1 for details.

Update has been a hot topic recently.  There's been discussion in [the meeting this week](https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2021-06-10) and on the slack channel, so you might be able to collaborate with others if you are just starting out on this.
